lib/web/desktop.go

```diff
@@ -87,34 +88,65 @@ func createDesktopConnection(
 
-	if len(winServices) == 0 {
-		return trace.NotFound("No windows_desktop_services are registered in this cluster")
-	}
 	// Pick a random Windows desktop service as our gateway.
 	// When agent mode is implemented in the service, we'll have to filter out
 	// the services in agent mode.
 	//
 	// In the future, we may want to do something smarter like latency-based
 	// routing.
-	service := winServices[rand.Intn(len(winServices))]
-	log = log.WithField("windows-service-uuid", service.GetName())
-	log = log.WithField("windows-service-addr", service.GetAddr())
-	serviceCon, err := site.DialTCP(reversetunnel.DialParams{
-		From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: r.RemoteAddr},
-		To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
-		ConnType: types.WindowsDesktopTunnel,
-		ServerID: service.GetName() + "." + ctx.parent.clusterName,
+	winDesktops, err := ctx.unsafeCachedAuthClient.GetWindowsDesktops(r.Context(),
+		types.WindowsDesktopFilter{Name: desktopName})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(winDesktops) == 0 {
+		return trace.NotFound("no windows_desktops were found")
+	}
+	var validServiceIDs []string
+	for _, desktop := range winDesktops {
+		if desktop.GetHostID() == "" {
+			// desktops with empty host ids are invalid and should
+			// only occur when migrating from an old version of teleport
+			continue
+		}
+		validServiceIDs = append(validServiceIDs, desktop.GetHostID())
+	}
+	rand.Shuffle(len(validServiceIDs), func(i, j int) {
+		validServiceIDs[i], validServiceIDs[j] = validServiceIDs[j], validServiceIDs[i]
 	})
+
+	var serviceCon net.Conn
+	for _, id := range validServiceIDs {
+		service, err := ctx.clt.GetWindowsDesktopService(context.Background(), id)
+		if err != nil {
+			log.Errorf("Error finding service with id %s", id)
+			continue
+		}
+		log = log.WithField("windows-service-uuid", service.GetName())
+		log = log.WithField("windows-service-addr", service.GetAddr())
+		serviceCon, err = site.DialTCP(reversetunnel.DialParams{
+			From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: r.RemoteAddr},
+			To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
+			ConnType: types.WindowsDesktopTunnel,
+			ServerID: service.GetName() + "." + ctx.parent.clusterName,
+		})
+		if err == nil {
+			break
+		}
+		if !trace.IsConnectionProblem(err) {
+			return trace.WrapWithMessage(err,
+				"error connecting to windows_desktop_service at %q: %v", service.GetAddr(), err)
+		}
+		log.Infof("Error connecting to service %q, trying another.", service.GetAddr())
+	}
 	if err != nil {
-		return trace.WrapWithMessage(err, "failed to connect to windows_desktop_service at %q: %v", service.GetAddr(), err)
+		return trace.Errorf("failed to connect to any windows_desktop_service: %v", err)
+	}
+	if serviceCon == nil {
+		return trace.Errorf("failed to connect to any windows_desktop_service: connection unexpectedly nil")
 	}
```

Comment:
nit: I like to see loops like this pulled into a function so you can just `return` rather than `break`ing out of the loop, and you don't even need the mutable `serviceCon` variable that you have to check for nil after the loop, but I won't block the PR for this